checker/src/org/checkerframework/checker/signedness/SignednessVisitor.java

```diff
@@ -149,6 +231,61 @@ private boolean isMaskedShift(BinaryTree shiftExpr) {
     }
 
     /**
+     * Determines if a right shift operation, {@code >>} or {@code >>>}, is type casted such that
+     * the cast renders the shift signedness ({@code >>} vs {@code >>>}) irrelevent by discarding
+     * the bits duplicated into the shift result. For example, the following pair of right shifts on
+     * {@code short s} both produce the same results under any input, because of type casting:
+     *
+     * <p>{@code (byte)(s >> 8) == (byte)(b >>> 8);}
+     *
+     * @param shiftExpr a right shift expression: {@code expr1 >> expr2} or {@code expr1 >>> expr2}
+     * @return true iff the right shift is type casted such that a signed or unsigned right shift
+     *     has the same effect
+     */
+    private boolean isCastedShift(BinaryTree shiftExpr) {
+        // enclosing is the operation or statement that immediately contains shiftExpr
+        Tree enclosing;
+        {
+            TreePath parentPath = visitorState.getPath().getParentPath();
+            enclosing = parentPath.getLeaf();
+            // Strip away all parentheses from the shift operation
+            while (enclosing.getKind() == Kind.PARENTHESIZED) {
+                parentPath = parentPath.getParentPath();
+                enclosing = parentPath.getLeaf();
+            }
+        }
+
+        if (!isPrimitiveCast(enclosing)) {
+            return false;
+        }
+
+        TypeCastTree cast = (TypeCastTree) enclosing;
+
+        // Determine the type of the cast target
+        Tree castType = cast.getType();
+        if (castType.getKind() != Kind.ANNOTATED_TYPE) {
+            return false;
+        }
+        AnnotatedTypeTree annotatedType = (AnnotatedTypeTree) castType;
+        ExpressionTree underlyingType = annotatedType.getUnderlyingType();
+        PrimitiveTypeTree castPrimitiveType = (PrimitiveTypeTree) underlyingType;
+        TypeKind castTypeKind = castPrimitiveType.getPrimitiveTypeKind();
+
+        // Determine the type of the shift result
+        TypeKind shiftTypeKind =
+                atypeFactory.getAnnotatedType(shiftExpr).getUnderlyingType().getKind();
+
+        // Determine shift literal
+        ExpressionTree shift = shiftExpr.getRightOperand();
```

Comment:
It's a bit confusing that you have three variables named `shift*`, and only this one has no suffix indicating its type.
I don't have a concrete suggestion, however, and this comment is more minor than other ones in this review.
(I suppose you could abstract it out into a routine that returns an Integer, or null if the shift amount is not known at compile time.)